### PR TITLE
Support finding "sibling" pages when using RedirecToPage

### LINKED ISSUE #6083: Support finding "sibling" pages when using RedirecToPage and other URL generation sites with relative page names
Consider a pages folder layout:
```
~/
  /Pages
    Index.cshtml
    /Account
      Login.cshtml
      Register.cshtml
```

To redirect from `Login.cshtml` to `Register.cshtml` a call like this is made: `return RedirectToPage("/Account/Register");`. This requires knowledge of the full path to the desired page from the configured pages root.

It would be nice to be able to redirect to pages in the same folder (sibling) without having to know the full page hierarchy like so: `return RedirectToPage("Register");`. Essentially, the page name passed here is assumed to relative to the current page if it's not qualified with a leading `/`. If there is a named page with the same name, the sibling page would take precedence over the named page.

This would make it possible to move hives of pages from one place to another and not have to update any of the URL-sensitive code or references within them.

Same logic would of course apply to anywhere we accept page names, e.g. `<a asp-page="Register">Click here to register</a>`

@rynowak @pranavkm

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #6083

### CODE REVIEW

**Comment by DamianEdwards** [comment]
Does this apply only to `RedirectToPage` calls or any place page URLs are generated?

**Comment by pranavkm** [comment]
I'll have to fix the tag helpers to call the `Page` method once it goes in. Trying to figure out the sequence in which this PR, https://github.com/aspnet/Mvc/pull/6180 and https://github.com/aspnet/Mvc/pull/6168 need to go in so this works

File: src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs [hunk 1]
```diff
@@ -440,5 +442,23 @@ public static string Page(this IUrlHelper urlHelper, string pageName)
...
         }
+
+        private static object CalculatePageName(ActionContext actionContext, string pageName)
+        {
+            Debug.Assert(pageName.Length > 0);
+            // Paths not qualified with a leading slash are treated as relative to the current page.
+            if (pageName[0] != '/')
+            {
+                var currentPagePath = NormalizedRouteValue.GetNormalizedRouteValue(actionContext, "page");
+                if (string.IsNullOrEmpty(currentPagePath))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: This should throw.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: This means that you're trying to use a page-relative feature from a non-page.

File: src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs [hunk 2]
```diff
@@ -440,5 +442,23 @@ public static string Page(this IUrlHelper urlHelper, string pageName)
                 host: host,
                 fragment: fragment);
         }
+
+        private static object CalculatePageName(ActionContext actionContext, string pageName)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: Why not unit test this directly?

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: We sort of already had the infrastructure to test the Page method directly. Seemed convenient to test it in entirety.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Core/Resources.resx [hunk 3]
```diff
@@ -403,4 +403,7 @@
   <data name="NoRoutesMatchedForPage" xml:space="preserve">
     <value>No page named '{0}' matches the supplied values.</value>
   </data>
+  <data name="UrlHelper_RelativePagePathIsNotSupported" xml:space="preserve">
+    <value>The relative page path '{0}' can only be from the context of executing a Razor page. Specify a root relative path with a leading '/' to calculate it in the current context.</value>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Resources.resx]: `page path` -> `page name`
`can only be from the context of executing a Razor page` -> `can only be used while executing a Razor Page`.
`to calculate it in the current context` -> `to generate a URL outside of a Razor Page`

**Pull request closed without merging**